include/aws/cal/symmetric_cipher.h

```diff
@@ -0,0 +1,263 @@
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_decrypt(
+    struct aws_symmetric_cipher *cipher,
+    const struct aws_byte_cursor *to_decrypt,
+    struct aws_byte_buf *out);
+
+/**
+ * Encrypts any remaining data that was reserved for final padding, loads GMACs etc... and if there is any
+ * writes any remaining encrypted data to out. If out is dynamic it will be expanded. If it is not, and
+ * out is not large enough to handle the decrypted output, the call will fail. If you're trying to optimize
+ *  to use a stack based array or something, make sure it's at least as large as the size of 2 BLOCKs to account for
+ * padding etc...
+ *
+ * After invoking this function, you MUST call aws_symmetric_cipher_reset() before invoking any encrypt/decrypt
+ * operations on this cipher again.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_finalize_encryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Decrypts any remaining data that was reserved for final padding, loads GMACs etc... and if there is any
+ * writes any remaining decrypted data to out. If out is dynamic it will be expanded. If it is not, and
+ * out is not large enough to handle the decrypted output, the call will fail. If you're trying to optimize
+ * to use a stack based array or something, make sure it's at least as large as the size of 2 BLOCKs to account for
+ * padding etc...
+ *
+ * After invoking this function, you MUST call aws_symmetric_cipher_reset() before invoking any encrypt/decrypt
+ * operations on this cipher again.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_finalize_decryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+
+/**
+ * Resets the cipher state for starting a new encrypt or decrypt operation. Note encrypt/decrypt cannot be mixed on the
+ * same cipher without a call to reset in between them. However, this leaves the key, iv etc... materials setup for
+ * immediate reuse.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_reset(struct aws_symmetric_cipher *cipher);
+
+/**
+ * Gets the current GMAC tag. If not AES GCM, this function will just copy an empty buffer over.
+ * You would typically call this function after calling aws_symmetric_cipher_finalize_encryption()
+ * to fetch the tag for transmitting it in a cryptographic protocol.
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_get_tag(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
```

Comment:
I'd be fine with this just returning a `struct aws_byte_cursor` by value.
yeah ... it's a bit memory unsafe, but it's so much simpler to use

imagine:
```
JNIEXPORT
jbyteArray JNICALL Java_software_amazon_awssdk_crt_cipher_get_tag(...) {
    ...
    bool success = false;
    struct aws_byte_buf tag_buf;
    aws_byte_buf_init(&tag_buf, aws_jni_get_allocator(), 256/*todo what's a good initial size?*/);
    if (aws_symmetric_cipher_get_tag(cipher)) {
        aws_jni_throw_runtime_exception(env, "failed to get tag");
        goto done;
    }

    success = true;
done:
    jbyteArray result = NULL;
    if (success) {
        struct aws_byte_cursor tag_cursor = aws_byte_cursor_from_buf(&tag_buf);
        result = aws_jni_byte_array_from_cursor(&tag_cursor);
    }
    aws_byte_buf_clean_up(&tag_buf);
    return result;
}
```

vs

```
JNIEXPORT
jbyteArray JNICALL Java_software_amazon_awssdk_crt_cipher_get_tag(...) {
    ...
    struct aws_byte_cursor tag = aws_symmetric_cipher_get_tag(cipher);
    return aws_jni_byte_array_from_cursor(&tag);
}```

Reply:
done